src/libraries/Common/src/Roslyn/MetadataTypeName.cs

```diff
@@ -0,0 +1,319 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#nullable disable
+
+using System;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis
+{
+    /// <summary>
+    /// Helper structure to encapsulate/cache various information about metadata name of a type and
+    /// name resolution options.
+    /// Also, allows us to stop using strings in the APIs that accept only metadata names,
+    /// making usage of them less bug prone.
+    /// </summary>
+    internal partial struct MetadataTypeName
+    {
+        /// <summary>
+        /// Full metadata name of a type, includes namespace name for top level types.
+        /// </summary>
+        private string _fullName;
+
+        /// <summary>
+        /// Namespace name for top level types.
+        /// </summary>
+        private string _namespaceName;
+
+        /// <summary>
+        /// Name of the type without namespace prefix, but possibly with generic arity mangling present.
+        /// </summary>
+        private string _typeName;
+
+        /// <summary>
+        /// Name of the type without namespace prefix and without generic arity mangling.
+        /// </summary>
+        private string _unmangledTypeName;
+
+        /// <summary>
+        /// Arity of the type inferred based on the name mangling. It doesn't have to match the actual
+        /// arity of the type.
+        /// </summary>
+        private short _inferredArity;
+
+        /// <summary>
+        /// While resolving the name, consider only types with this arity.
+        /// (-1) means allow any arity.
+        /// If forcedArity >= 0 and useCLSCompliantNameArityEncoding, lookup may
+        /// fail because forcedArity doesn't match the one encoded in the name.
+        /// </summary>
+        private short _forcedArity;
+
+        /// <summary>
+        /// While resolving the name, consider only types following
+        /// CLS-compliant generic type names and arity encoding (ECMA-335, section 10.7.2).
+        /// I.e. arity is inferred from the name and matching type must have the same
+        /// emitted name and arity.
+        /// TODO: PERF: Encode this field elsewhere to save 4 bytes
+        /// </summary>
+        private bool _useCLSCompliantNameArityEncoding;
+
+#if false
```

Comment:
similarly, this type is heavily if'defed out.

Reply:
@stephentoub do you know if we are also running the linker on the source generators? If so, we probably can keep this code (without the ifdef) and then just have the linker strip it out, and that way we can still preserve the files in sync with Roslyn, while not having to carry the extra IL in our binaries.